### DISCUSSION TITLE
Pagination design help

### DISCUSSION THREAD

**Discussion by bcthomas** [falconry/falcon#2030] [Q&A]
I have a nice little falcon-based API in the works, tied up to a database using peewee. I have one model that has thousands of children and I'd like to paginate the responses for it.

Does anyone have any example code they could share on implementing pagination in falcon or suggestions for a proper design?

**Comment by swistakm** [comment]
I feel really bad about advertising my own projects in other project's issue threads but you can take a look at [graceful](https://github.com/swistakm/graceful). It's a toolkit for building resources for falcon. It can be used even if you do not want to use it for every resource in your code.

It already comes with two predefined parameters for pagination (`page` and `page_size`) but you can easily add new ones for other types of pagination flows (like limit/offset or cursor). Even if this will not fit your project take a look on how query params are handled/defined in this project. I hope you find it inspiring in solving your problem.

**Comment by yohanboniface** [comment]
@bcthomas I've come up with this very simple solution https://github.com/etalab/ban/blob/master/ban/http/resources.py#L45
But I plan to implement cursor based pagination in a later step (see http://use-the-index-luke.com/no-offset)

**Comment by bcthomas** [comment]
@swistakm thanks. interesting and well designed toolkit.  I will have a look.

@yohanboniface I really like the idea of cursor-based pagination. I will be interested in how you implement. I read a good bit about the disqus cursor system in their API. Seems like a good strategy.

**Comment by Gesias** [comment]
:+1: for this one, @yohanboniface your BaseCollection thing will have to do for now thanks for the tip!

**Comment by vytas7** [comment]
I'm going to convert this issue to a discussion, since it's focused on a specific area in which Falcon doesn't come with batteries included. Documenting general ORM integration is tracked in https://github.com/falconry/falcon/issues/1407.
